runtime/memprof.c

```diff
@@ -15,6 +15,276 @@
+   * used a doubly-linked list, but that's premature optimisation
+   * at this point. */
+  memprof_thread_t *p = &domain->threads;
+  while (*p != thread) {
+    p = &(*p)->next;
+  }
+
+  *p = thread->next;
+
+  caml_stat_free(thread);
+}
+
+/**** Create and destroy domain state structures ****/
+
+static void domain_destroy(memprof_domain_t domain)
+{
+  memprof_thread_t thread = domain->threads;
+  while (thread) {
+    memprof_thread_t next = thread->next;
+    thread_destroy(thread);
+    thread = next;
+  }
+
+  caml_stat_free(domain);
+}
+
+static memprof_domain_t domain_create(caml_domain_state *caml_state)
+{
+  memprof_domain_t domain = caml_stat_alloc(sizeof(memprof_domain_s));
+  if (!domain) {
+    return NULL;
+  }
+
+  domain->caml_state = caml_state;
+  domain->threads = NULL;
+  domain->current = NULL;
+  atomic_store(&domain->config.stopped, true);
+  domain->config.lambda = 0.0;
+  domain->config.one_log1m_lambda = 0.0;
+  domain->config.callstack_size = 0;
+  domain->config.tracker = Val_unit;
+
+  /* create initial thread for domain */
+  memprof_thread_t thread = thread_create(domain);
+  if (thread) {
+    domain->current = thread;
+  } else {
+    domain_destroy(domain);
+    domain = NULL;
+  }
+  return domain;
+}
+
+/**** Interface to domain module ***/
+
+void caml_memprof_new_domain(caml_domain_state *parent,
+                             caml_domain_state *child)
+{
+  memprof_domain_t domain = domain_create(child);
+
+  child->memprof = domain;
+  /* if parent domain is profiling, child domain should also be profiling */
+  if (domain && parent && parent->memprof
+      && !atomic_load(&parent->memprof->config.stopped)) {
+    domain->config = parent->memprof->config;
+  }
+}
+
+void caml_memprof_delete_domain(caml_domain_state *domain)
+{
+  if (!domain->memprof) {
+    return;
+  }
+  domain_destroy(domain->memprof);
+  domain->memprof = NULL;
+}
+
+/**** Interface with domain action-pending flag ****/
+
+/* If profiling is active in the current domain, and we may have some
+ * callbacks pending, set the action pending flag. */
+
+static void set_action_pending_as_needed(memprof_domain_t domain)
+{
+  /* if (condition) caml_set_action_pending(domain->caml_state); */
+}
+
+/* Set the suspended flag on `domain` to `s`. */
+
+static void update_suspended(memprof_domain_t domain, bool s)
+{
+  if (domain->current) {
+    domain->current->suspended = s;
+  }
+  caml_memprof_renew_minor_sample(domain->caml_state);
+  if (!s) set_action_pending_as_needed(domain);
+}
+
+/* Set the suspended flag on the current domain to `s`. */
+
+void caml_memprof_update_suspended(bool s) {
+  update_suspended(Caml_state->memprof, s);
+}
+
+/**** Sampling procedures ****/
+
+Caml_inline bool running(memprof_domain_t domain)
+{
+  memprof_thread_t thread = domain->current;
+
+  return (thread
+          && !atomic_load(&domain->config.stopped)
+          && domain->config.lambda > 0
```

Comment:
This is an excellent question. This is descended from a similar test in the old statmemprof. I think probably we should raise an exception in `start` if lambda == 0 (as we do for lambda < 0 or lambda > 1). In the old statmemprof, starting with zero lambda does not raise an exception but (a) all sampling is short-circuited off (rather than simply never occurring), and (b) stopping is done, in part, by setting lambda to zero. New statmemprof has more explicit stopping mechanisms (and more complex, due to multicore), so using lambda for this could go away.

Reply:
Ok, let's do that, then.